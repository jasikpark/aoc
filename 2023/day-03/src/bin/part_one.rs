fn main() {
    todo!()
}

pub fn part_one(_input: &str) -> usize {
    todo!()
}

mod test {
    #[test]
    fn test_example() {
        use crate::part_one;
        let input = r#"467..114..
        ...*......
        ..35..633.
        ......#...
        617*......
        .....+.58.
        ..592.....
        ......755.
        ...$.*....
        .664.598.."#;
        assert_eq!(part_one(input), 4361);
    }
}
