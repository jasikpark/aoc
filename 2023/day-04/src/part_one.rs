use nom::{
    bytes::complete::tag,
    character::complete::digit1,
    combinator::map_res,
    multi::{many1, separated_list1},
    sequence::tuple,
    IResult,
};

fn digit_parser(input: &str) -> IResult<&str, usize> {
    map_res(digit1, str::parse)(input)
}

pub struct ScratchoffCard {
    card_num: usize,
    /// The first correct number scores 1 point, then each after that doubles the score.
    score: usize,
    winning_numbers: LotteryNums,
    my_numbers: LotteryNums,
}

impl ScratchoffCard {
    fn parse(input: &str) -> IResult<&str, Self> {
        let (input, _) = tuple((tag("Card"), many1(tag(" "))))(input)?;
        let (input, card_num) = digit_parser(input)?;
        let (input, _) = tag(": ")(input)?;
        let (input, lottery_numbers) = separated_list1(
            tuple((many1(tag(" ")), tag("|"), many1(tag(" ")))),
            parse_lottery_numbers,
        )(input)?;

        assert_eq!(lottery_numbers.len(), 2);

        let winning_numbers = lottery_numbers[0].clone();
        let my_numbers = lottery_numbers[1].clone();

        let my_winning_numbers_count = my_numbers
            .iter()
            .filter(|num| winning_numbers.contains(num))
            .count();

        let score = Self::calc_score(my_winning_numbers_count);

        Ok((
            input,
            Self {
                card_num,
                winning_numbers,
                my_numbers,
                score,
            },
        ))
    }

    /// ```
    /// use day_4::part_one::ScratchoffCard;
    /// assert_eq!(ScratchoffCard::calc_score(0), 0);
    /// assert_eq!(ScratchoffCard::calc_score(1), 1);
    /// assert_eq!(ScratchoffCard::calc_score(2), 2);
    /// assert_eq!(ScratchoffCard::calc_score(3), 4);
    /// assert_eq!(ScratchoffCard::calc_score(4), 8);
    /// ```
    pub fn calc_score(winning_numbers: usize) -> usize {
        if winning_numbers == 0 {
            return 0;
        }
        if winning_numbers == 1 {
            return 1;
        }
        let mut score = 1;
        // Double the score for every additional winning number.
        for _ in 1..winning_numbers {
            score *= 2;
        }

        score
    }
}

type LotteryNums = Vec<usize>;

/// ```
/// use day_4::part_one::parse_lottery_numbers;
/// let input = "1 2 3 14 69 11";
/// assert_eq!(parse_lottery_numbers(input).unwrap(), ("", vec![1, 2, 3, 14, 69, 11]));
/// ```
pub fn parse_lottery_numbers(input: &str) -> IResult<&str, LotteryNums> {
    separated_list1(many1(tag(" ")), digit_parser)(input)
}

pub fn part_one(input: &str) -> usize {
    input
        .lines()
        .flat_map(|line| ScratchoffCard::parse(line).map(|card| card.1.score))
        .sum()
}

mod test {
    #[test]
    fn provided_testcase() {
        use crate::part_one::part_one;
        let input = r#"Card 1: 41 48 83 86 17 | 83 86  6 31 17  9 48 53
        Card 2: 13 32 20 16 61 | 61 30 68 82 17 32 24 19
        Card 3:  1 21 53 59 44 | 69 82 63 72 16 21 14  1
        Card 4: 41 92 73 84 69 | 59 84 76 51 58  5 54 83
        Card 5: 87 83 26 28 32 | 88 30 70 12 93 22 82 36
        Card 6: 31 18 13 56 72 | 74 77 10 23 35 67 36 11"#;
        assert_eq!(part_one(input), 8);
    }

    #[test]
    fn provided_input() {
        use crate::part_one::part_one;
        let input = include_str!("./input.txt");
        assert_eq!(part_one(input), 8);
    }
}
