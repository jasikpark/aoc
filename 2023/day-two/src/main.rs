use nom::{
    bytes::complete::tag,
    character::complete::{alpha1, digit1},
    combinator::{map, map_res},
    multi::separated_list1,
    sequence::separated_pair,
    IResult,
};

fn main() {
    dbg!(Game::parse("Game 1: 3 blue, 4 red;").unwrap());
}

pub fn part_one(input: &str) -> usize {
    let games = input
        .lines()
        .map(|line| {
            let (input, game) = Game::parse(line.trim()).unwrap();
            assert!(input.is_empty());
            game
        })
        .collect::<Vec<_>>();

    dbg!(games);

    todo!();
}

#[derive(Debug, PartialEq)]
enum Color {
    Red,
    Blue,
    Green,
}

#[derive(Debug, PartialEq)]
struct ColorCount {
    pub count: usize,
    pub color: Color,
}

#[derive(Debug, PartialEq)]
struct Hand {
    pub color_counts: Vec<ColorCount>,
}

impl Hand {
    fn parse(input: &str) -> IResult<&str, Self> {
        let color_count_parser = map(Self::color_count_parser, |(count, color)| ColorCount {
            count,
            color: match color {
                "red" => Color::Red,
                "blue" => Color::Blue,
                "green" => Color::Green,
                _ => panic!("pls don't have other colors"),
            },
        });

        let (input, color_counts) = separated_list1(tag(", "), color_count_parser)(input)?;

        Ok((input, Self { color_counts }))
    }

    fn color_count_parser(input: &str) -> IResult<&str, (usize, &str)> {
        separated_pair(digit_parser, tag(" "), alpha1)(input)
    }
}

#[derive(Debug, PartialEq)]
struct Game {
    pub game_num: usize,
    pub hands: Vec<Hand>,
}

fn digit_parser(input: &str) -> IResult<&str, usize> {
    map_res(digit1, str::parse)(input)
}

impl Game {
    fn parse(input: &str) -> IResult<&str, Self> {
        let (input, _) = tag("Game ")(input)?;
        let (input, game_num) = digit_parser(input)?;
        let (input, _) = tag(": ")(input)?;

        let mut game_parser = separated_list1(tag("; "), Hand::parse);

        let (input, hands) = game_parser(input)?;

        Ok((input, Self { game_num, hands }))
    }
}

mod test {

    #[test]
    fn initial_example_part_one() {
        use crate::part_one;
        let input = r#"Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
        Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue
        Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red
        Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red
        Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green"#;
        assert_eq!(part_one(input), 8);
    }

    #[test]
    fn test_example_part_one() {
        use crate::part_one;
        let input = include_str!("./input.txt");
        assert_eq!(part_one(input), 8);
    }

    #[test]
    fn snapshot_game_parse() {
        use crate::Game;

        let input_1 = "Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green";
        let input_2 = "Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue";
        let input_3 = "Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red";
        let input_4 = "Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red";
        let input_5 = "Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green";

        insta::assert_debug_snapshot!(Game::parse(input_1));
        insta::assert_debug_snapshot!(Game::parse(input_2));
        insta::assert_debug_snapshot!(Game::parse(input_3));
        insta::assert_debug_snapshot!(Game::parse(input_4));
        insta::assert_debug_snapshot!(Game::parse(input_5));
    }
}
