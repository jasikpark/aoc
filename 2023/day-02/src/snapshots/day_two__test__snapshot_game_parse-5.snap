---
source: day-two/src/main.rs
expression: "Game::parse(input_5)"
---
Ok(
    (
        "",
        Game {
            game_num: 5,
            hands: [
                Hand {
                    color_counts: [
                        ColorCount {
                            count: 6,
                            color: Red,
                        },
                        ColorCount {
                            count: 1,
                            color: Blue,
                        },
                        ColorCount {
                            count: 3,
                            color: Green,
                        },
                    ],
                },
                Hand {
                    color_counts: [
                        ColorCount {
                            count: 2,
                            color: Blue,
                        },
                        ColorCount {
                            count: 1,
                            color: Red,
                        },
                        ColorCount {
                            count: 2,
                            color: Green,
                        },
                    ],
                },
            ],
        },
    ),
)
